[export][cond] support merging constant ints as unbacked symint

ydwu4 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #152742


@pianpwk points out that this will be helpful to address several data dependent issues in huggingface [models](https://github.com/huggingface/diffusers/blob/e23705e5577387872dd55ebf6db81bd59df928f1/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py#L332) with the following pattern:
```python
idx = return 0 if u0 else return 1
return  x[idx]
```
We could preserve the conditional with a cond.


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

ydwu4 committed:
[export][cond] support merging constant ints as unbacked symint

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][cond] support merging constant ints as unbacked symint"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][cond] support merging constant ints as unbacked symint"



pianpwk points out that this will be helpful to address several data dependent issues in huggingface [models](https://github.com/huggingface/diffusers/blob/e23705e5577387872dd55ebf6db81bd59df928f1/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py#L332) with the following pattern:
```python
idx = if u0 return 0 else return 1
return  x[idx]
```
We could preserve the conditional with a cond.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][cond] support merging constant ints as unbacked symint"



pianpwk points out that this will be helpful to address several data dependent issues in huggingface [models](https://github.com/huggingface/diffusers/blob/e23705e5577387872dd55ebf6db81bd59df928f1/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py#L332) with the following pattern:
```python
idx = if u0 return 0 else return 1
return  x[idx]
```
We could preserve the conditional with a cond.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 commented on torch/_export/serde/serialize.py:
```diff
@@ -1626,11 +1626,13 @@ def save_graph_module(self) -> Iterator[None]:
             self.module,
             self.serialized_name_to_node,
             self.serialized_name_to_meta,
+            self.unbacked_symbols
```
We need stash unbacked_symbols before deserialize_subgraph because the subgraph may have different set of the unbacked_symbols.

ydwu4 committed:
Update on "[export][cond] support merging constant ints as unbacked symint"



pianpwk points out that this will be helpful to address several data dependent issues in huggingface [models](https://github.com/huggingface/diffusers/blob/e23705e5577387872dd55ebf6db81bd59df928f1/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py#L332) with the following pattern:
```python
idx = if u0 return 0 else return 1
return  x[idx]
```
We could preserve the conditional with a cond.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

zou3519 commented on test/export/test_export.py:
```diff
@@ -1299,6 +1299,17 @@ def false_fn(x):
+        m = M()
+        ep = export(M(), args)
+        self.assertEqual(m(*args), ep.module()(*args))
+
```
Can we get a self.assertEquals on the graph? I'm not convinced that this doesn't just bake a branch in the graph lol

ydwu4 replied:
sure lol, added the graph check.

ydwu4 committed:
Update on "[export][cond] support merging constant ints as unbacked symint"



pianpwk points out that this will be helpful to address several data dependent issues in huggingface [models](https://github.com/huggingface/diffusers/blob/e23705e5577387872dd55ebf6db81bd59df928f1/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py#L332) with the following pattern:
```python
idx = if u0 return 0 else return 1
return  x[idx]
```
We could preserve the conditional with a cond.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 commented on torch/_higher_order_ops/cond.py:
```diff
@@ -458,6 +460,26 @@ def _merge_tensors(
+            _bound(s0, False), _bound(s1, False)
+        )
+
+    if type(a) is int and type(b) is int:
+        assert mode.shape_env is not None
+        merged_out = mode.shape_env.create_unbacked_symint()
+        mode.shape_env.constrain_symbol_range(merged_out.node.expr, *min_max(a, b))
+        return merged_out
```
yes, that's an optimization we can do though seems rare in practice. I'll add a test for this case.

ydwu4 committed:
Update on "[export][cond] support merging constant ints as unbacked symint"



pianpwk points out that this will be helpful to address several data dependent issues in huggingface [models](https://github.com/huggingface/diffusers/blob/e23705e5577387872dd55ebf6db81bd59df928f1/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py#L332) with the following pattern:
```python
idx = if u0 return 0 else return 1
return  x[idx]
```
We could preserve the conditional with a cond.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][cond] support merging constant ints as unbacked symint"



pianpwk points out that this will be helpful to address several data dependent issues in huggingface [models](https://github.com/huggingface/diffusers/blob/e23705e5577387872dd55ebf6db81bd59df928f1/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py#L332) with the following pattern:
```python
idx = if u0 return 0 else return 1
return  x[idx]
```
We could preserve the conditional with a cond.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 committed:
Update on "[export][cond] support merging constant ints as unbacked symint"



pianpwk points out that this will be helpful to address several data dependent issues in huggingface [models](https://github.com/huggingface/diffusers/blob/e23705e5577387872dd55ebf6db81bd59df928f1/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py#L332) with the following pattern:
```python
idx = if u0 return 0 else return 1
return  x[idx]
```
We could preserve the conditional with a cond.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot revert -m "breaking trunk"

ydwu4 commented:
@pytorchbot revert -m "breaking trunk" -c "nosignal"

ydwu4 committed:
Update on "[export][cond] support merging constant ints as unbacked symint"



pianpwk points out that this will be helpful to address several data dependent issues in huggingface [models](https://github.com/huggingface/diffusers/blob/e23705e5577387872dd55ebf6db81bd59df928f1/src/diffusers/schedulers/scheduling_euler_ancestral_discrete.py#L332) with the following pattern:
```python
idx = return 0 if u0 else return 1
return  x[idx]
```
We could preserve the conditional with a cond.


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames

[ghstack-poisoned]

ydwu4 commented:
@pytorchbot merge

ydwu4 commented:
@pytorchbot merge
